File: server/repository/minio_repository.py
from datetime import datetime
from urllib.parse import urlparse

from s3 import minio_client
from server.enums.FigType import FigType

BUCKET_NAME = "plot-figure"


class MinioRepository:
    @staticmethod
    def save_image(fig_type: FigType, image_buf, file_name):
        ensure_bucket_exists(BUCKET_NAME)

        now = str(datetime.today()).replace(" ", "-")
        name_path = f"images/{fig_type.value}/{file_name}_{fig_type.value}_{now}"
        if fig_type in fig_type.image_fig_type_list:
            obj_name = name_path + ".png"
        else:
            obj_name = name_path + ".gif"

        minio_client.put_object(
            BUCKET_NAME,
            obj_name,
            image_buf,
            length=image_buf.getbuffer().nbytes,
            content_type="image/png",
        )

        image_url = f"http://localhost:9000/{BUCKET_NAME}/{obj_name}"
        return image_url

    # ファイルを削除する関数
    @staticmethod
    def delete_file(url):
        bucket_name, object_name = extract_bucket_and_object(url)
        try:
            minio_client.remove_object(bucket_name, object_name)
        except Exception as e:
            print(f"Error removing object: {e}")

    # バケット内の指定されたディレクトリの全てのオブジェクトを削除する関数
    @staticmethod
    def delete_objects_in_directory(directory):
        try:
            objects = minio_client.list_objects(
                BUCKET_NAME, prefix=directory, recursive=True
            )
            for obj in objects:
                minio_client.remove_object(BUCKET_NAME, obj.object_name)
        except Exception as e:
            print(f"Error deleting objects: {e}")


def ensure_bucket_exists(bucket_name):
    if not minio_client.bucket_exists(bucket_name):
        minio_client.make_bucket(bucket_name)


def extract_bucket_and_object(url):
    parsed_url = urlparse(url)
    if parsed_url.scheme != "http" and parsed_url.scheme != "https":
        raise ValueError(
            "Invalid URL scheme. Only 'http' and 'https' schemes are supported."
        )

    parts = parsed_url.path.split("/", 2)
    if len(parts) < 3:
        raise ValueError("Invalid URL format. Bucket name and object name are missing.")

    bucket_name = parts[1]
    object_name = parts[2]

    return bucket_name, object_name

File: go-backend/repository/minioRepository.go
package repository

import (
	"bytes"
	"context"
	"fmt"
	"log"
	"net/url"
	"strings"
	"time"

	"github.com/kkito0726/mea-viewer/db"
	"github.com/kkito0726/mea-viewer/model"
	"github.com/minio/minio-go/v7"
)

const BUCKET_NAME = "plot-figure"

func SaveImage(imageBuf *bytes.Buffer, formDto *model.FormDto) (string, error) {
	ensureBucketExists(BUCKET_NAME)

	now := time.Now().Format("2006-01-02-15-04-05")
	objName := fmt.Sprintf("images/%s/%d_%s_%s_%s.png", formDto.FigType.String(), formDto.Ch, formDto.FileName, formDto.FigType.String(), now)

	_, err := db.MinioClient.PutObject(
		context.Background(),
		BUCKET_NAME,
		objName,
		imageBuf,
		int64(imageBuf.Len()),
		minio.PutObjectOptions{ContentType: "image/png"},
	)
	if err != nil {
		return "", err
	}

	imageURL := fmt.Sprintf("http://localhost:9000/%s/%s", BUCKET_NAME, objName)
	return imageURL, nil
}

// DeleteFile はMinIOからファイルを削除する関数
func DeleteFile(urlStr string) error {
	minioModel, err := ExtractBucketAndObject(urlStr)
	if err != nil {
		return err
	}

	err = db.MinioClient.RemoveObject(context.Background(), minioModel.BucketName, minioModel.ObjectName, minio.RemoveObjectOptions{})
	if err != nil {
		return fmt.Errorf("error removing object: %v", err)
	}
	return nil
}

func DeleteObjectsInDirectory(directory string) error {
	objectCh := db.MinioClient.ListObjects(context.Background(), BUCKET_NAME, minio.ListObjectsOptions{
		Prefix:    directory,
		Recursive: true,
	})

	for obj := range objectCh {
		if obj.Err != nil {
			return fmt.Errorf("error listing objects: %v", obj.Err)
		}
		err := db.MinioClient.RemoveObject(context.Background(), BUCKET_NAME, obj.Key, minio.RemoveObjectOptions{})
		if err != nil {
			return fmt.Errorf("error deleting objects: %v", err)
		}
	}
	return nil
}

func ensureBucketExists(bucketName string) {
	exists, err := db.MinioClient.BucketExists(context.Background(), bucketName)
	if err != nil {
		log.Fatalln(err)
	}
	if !exists {
		err = db.MinioClient.MakeBucket(context.Background(), bucketName, minio.MakeBucketOptions{})
		if err != nil {
			log.Fatalln(err)
		}
	}
}

// ExtractBucketAndObject はURLからバケット名とオブジェクト名を抽出する関数
func ExtractBucketAndObject(urlStr string) (model.MinioModel, error) {
	parsedURL, err := url.Parse(urlStr)
	if err != nil {
		return model.MinioModel{}, fmt.Errorf("invalid URL: %v", err)
	}

	parts := parsedURL.Path[1:]
	splitParts := strings.SplitN(parts, "/", 2)
	if len(splitParts) < 2 {
		return model.MinioModel{}, fmt.Errorf("invalid URL format. Bucket name and object name are missing")
	}

	return model.MinioModel{
		BucketName: splitParts[0],
		ObjectName: splitParts[1],
	}, nil
}
